docs/readmes/proposals/p018_control_network_metrics.md

```diff
@@ -392,6 +394,28 @@ since destination IP addresses and ports have very large domains, the overhead
 of having separate time-series would be too high. We instead decided to label
 Magma traffic by destination cloud service.
 
+### Performance impact of L2 instrumentation
+
+We evaluated the performance impact of instrumenting the `net_dev_start_xmit`
+kernel event using the Spirent lab setup as we did for [`tcp_sendmsg`](#performance). Two
+tests were run, one with `net_dev_start_xmit` instrumentation disabled, and one
+with it enabled. Each test had 600 user devices, 5 MB download rate per device,
+and an attach rate of 10. The tests ran for 30 minutes each.
+
+When `net_dev_start_xmit` instrumentation was disabled, the mean download
+throughput was roughly 540 Mbps. When it was enabled, the mean download
+throughput reduced to 430 Mbps, a 20% reduction.
+
+We also stress-tested the `net_dev_start_xmit` instrumentation. The most
+expensive operation in the relevant callback is incrementing a value in the BPF
+hash map. We put the increment operation in a loop and iterated it `n` times.
+When `n = 100`, the mean download throughput reduction was modest: from 430 Mbps
+to 420 Mbps. For `n = 150` and `n = 175`, no noticeable reduction was observed
+compared to `n = 100`. However, at `n = 200`, download throughput collapsed
+completely and produced a mean of around 5.5 Mbps with the peak hitting 75 Mbps.
+This sudden drop in throughput may have to do with excessive packet loss because
+of the slow callback function.
+
```

Comment:
can you also add similar results for `tcp_sendmsg` hook based BPF implementation.

Reply:
Those results are already present in this file: https://sourcegraph.com/github.com/magma/magma/-/blob/docs/readmes/proposals/p018_control_network_metrics.md?L285&view=code